frontends/election-manager/src/components/format_usb_modal.tsx

```diff
@@ -0,0 +1,178 @@
+import React, { useCallback, useContext, useEffect, useState } from 'react';
+import { assert, throwIllegalValue } from '@votingworks/utils';
+
+import {
+  Button,
+  Modal,
+  Prose,
+  isElectionManagerAuth,
+  isSystemAdministratorAuth,
+  LinkButton,
+  Loading,
+  UsbImage,
+} from '@votingworks/ui';
+import { AppContext } from '../contexts/app_context';
+
+export interface FormatUsbModalProps {
+  onClose: () => void;
+}
+
+type ModalState = 'init' | 'confirm' | 'formatting' | 'done' | 'error';
+
+export function FormatUsbModal({ onClose }: FormatUsbModalProps): JSX.Element {
+  const { usbDrive, auth } = useContext(AppContext);
+  assert(isSystemAdministratorAuth(auth) || isElectionManagerAuth(auth));
+  const userRole = auth.user.role;
+
+  const [state, setState] = useState<ModalState>('init');
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const format = useCallback(async () => {
+    setState('formatting');
+    try {
+      await usbDrive.format(userRole, { action: 'eject' });
+      setState('done');
+    } catch (error) {
+      setState('error');
+      setErrorMessage((error as Error).message);
+    }
+  }, [usbDrive, userRole]);
+
+  // Handle USB drive being removed during stages of the  process
+  useEffect(() => {
+    if (usbDrive.status === 'absent') {
+      setState('init');
+    }
+  }, [state, usbDrive.status]);
```

Comment:
To avoid needing this effect, I might put the formatting flow and its associated state in a sub-component that only renders when the usb drive is mounted.

(See [docs](https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) for more on why adjusting state based on props is considered an anti-pattern)

Reply:
Changed.